// vite_react_shadcn_ts/vite.config.ts.timestamp-1755429282212-fd304e8e976ae.mjs
// vite.config.ts
import { defineConfig } from "file:///C:/Users/Administrator/Sehat%20Kor/FullStack-SehatKor-Website-ash-cloud/client/node_modules/vite/dist/node/index.js";
import react from "file:///C:/Users/Administrator/Sehat%20Kor/FullStack-SehatKor-Website-ash-cloud/client/node_modules/@vitejs/plugin-react-swc/index.js";
import legacy from "file:///C:/Users/Administrator/Sehat%20Kor/FullStack-SehatKor-Website-ash-cloud/client/node_modules/@vitejs/plugin-legacy/dist/index.mjs";
import path from "path";
import { componentTagger } from "file:///C:/Users/Administrator/Sehat%20Kor/FullStack-SehatKor-Website-ash-cloud/client/node_modules/lovable-tagger/dist/index.js";
var __vite_injected_original_dirname = "C:\\Users\\Administrator\\Sehat Kor\\FullStack-SehatKor-Website-ash-cloud\\client";
var vite_config_default = defineConfig(({ mode }) => ({
  server: {
    host: "0.0.0.0",
    port: 8080,
    hmr: {
      host: "localhost",
      protocol: "ws",
      port: 8080
    }
  },
  plugins: [
    react(),
    legacy({
      // Target a broad set of browsers including older Safari/Firefox/Edge
      targets: [
        "defaults",
        "not dead",
        "iOS >= 12",
        "Safari >= 12",
        "Firefox >= 68",
        "Edge >= 79"
      ],
      additionalLegacyPolyfills: ["regenerator-runtime/runtime"],
      renderLegacyChunks: true,
      modernPolyfills: true
    }),
    mode === "development" && componentTagger()
  ].filter(Boolean),
  resolve: {
    alias: {
      "@": path.resolve(__vite_injected_original_dirname, "./src")
    }
  },
  esbuild: {
    target: "es2017"
  },
  build: {
    // Transpile down to broadly supported syntax
    target: "es2017",
    cssTarget: ["chrome61", "safari11", "firefox60"]
  }
}));
export {
  vite_config_default as default
};
